# Préparation des données pour un nettoyage manuel
Dans ce notebook nous allons filter les données afin des les réduire à une quantité nettoyable manuellement.

In [1]:
import pandas as pd
import os
import csv
from collections import Counter

from constants import *

%load_ext autoreload
%autoreload 2

## Regroupe toutes les données en un fichier

In [2]:
files = os.listdir(POPULATION_DATA_DIR)
census_files = [file for file in files if '.xlsx' in file]
census_files = sorted(census_files)
census_files

['1835.xlsx',
 '1836.xlsx',
 '1837.xlsx',
 '1838.xlsx',
 '1839.xlsx',
 '1840.xlsx',
 '1841.xlsx',
 '1842.xlsx',
 '1843.xlsx',
 '1844.xlsx',
 '1845.xlsx',
 '1846.xlsx',
 '1847.xlsx',
 '1848.xlsx',
 '1849.xlsx',
 '1850.xlsx',
 '1851.xlsx',
 '1852.xlsx',
 '1853.xlsx',
 '1854.xlsx',
 '1855.xlsx',
 '1856.xlsx',
 '1857.xlsx',
 '1858.xlsx',
 '1859.xlsx',
 '1860.xlsx',
 '1861.xlsx',
 '1862.xlsx',
 '1863.xlsx',
 '1864.xlsx',
 '1865.xlsx',
 '1866.xlsx',
 '1867.xlsx',
 '1868.xlsx',
 '1869.xlsx',
 '1870.xlsx',
 '1871.xlsx',
 '1872.xlsx',
 '1873.xlsx',
 '1874.xlsx',
 '1875.xlsx',
 '1876.xlsx',
 '1877.xlsx',
 '1878.xlsx',
 '1879.xlsx',
 '1880.xlsx',
 '1881.xlsx',
 '1882.xlsx',
 '1883.xlsx',
 '1884.xlsx',
 '1885.xlsx',
 '1886.xlsx',
 '1887.xlsx',
 '1888.xlsx',
 '1889.xlsx',
 '1890.xlsx',
 '1891.xlsx',
 '1892.xlsx',
 '1893.xlsx',
 '1894.xlsx',
 '1895.xlsx',
 '1896.xlsx',
 '1897.xlsx',
 '1898.xlsx']

In [3]:
df_example = pd.read_excel(POPULATION_DATA_DIR+'1861.xlsx', index_col=0)

df_raw = pd.DataFrame(columns=df_example.columns)
for filename in census_files:
    print(f"Read file: {filename}")
    df_year = pd.read_excel(POPULATION_DATA_DIR+filename, index_col=0)
    df_year['annee'] = filename.replace('.xlsx','').replace('_', '')
    df_raw = df_raw.append(df_year)
print(len(df_raw))

Read file: 1835.xlsx
Read file: 1836.xlsx
Read file: 1837.xlsx
Read file: 1838.xlsx
Read file: 1839.xlsx
Read file: 1840.xlsx
Read file: 1841.xlsx
Read file: 1842.xlsx
Read file: 1843.xlsx
Read file: 1844.xlsx
Read file: 1845.xlsx
Read file: 1846.xlsx
Read file: 1847.xlsx
Read file: 1848.xlsx
Read file: 1849.xlsx
Read file: 1850.xlsx
Read file: 1851.xlsx
Read file: 1852.xlsx
Read file: 1853.xlsx
Read file: 1854.xlsx
Read file: 1855.xlsx
Read file: 1856.xlsx
Read file: 1857.xlsx
Read file: 1858.xlsx
Read file: 1859.xlsx
Read file: 1860.xlsx
Read file: 1861.xlsx
Read file: 1862.xlsx
Read file: 1863.xlsx
Read file: 1864.xlsx
Read file: 1865.xlsx
Read file: 1866.xlsx
Read file: 1867.xlsx
Read file: 1868.xlsx
Read file: 1869.xlsx
Read file: 1870.xlsx
Read file: 1871.xlsx
Read file: 1872.xlsx
Read file: 1873.xlsx
Read file: 1874.xlsx
Read file: 1875.xlsx
Read file: 1876.xlsx
Read file: 1877.xlsx
Read file: 1878.xlsx
Read file: 1879.xlsx
Read file: 1880.xlsx
Read file: 1881.xlsx
Read file: 18

In [4]:
print(df_raw.columns)
df_jobs = df_raw[['chef_vocation','enfants_profession','pensionnaires_condition','annee']].copy()
df_jobs.to_csv(COMPLETE_RAW_DATA,index=False)

Index(['nom_rue', 'no_maison', 'proprietaire_nom', 'chef_prenom', 'chef_nom',
       'chef_annee_naissance', 'epouse_nom', 'epouse_annee_naissance',
       'enfants_dans_la_commune_prenom', 'enfants_annee_naissance',
       'chef_origine', 'chef_vocation', 'chef_recepisse',
       'pensionnaires_prenom', 'pensionnaires_nom',
       'pensionnaires_annee_naissance', 'pensionnaires_origine',
       'pensionnaires_condition', 'pensionnaires_recepisse', 'observations',
       'Page', 'chef_annee_arrivee', 'annee', 'nom_rue_corr',
       'enfants_chez_parents_prenom', 'epouse_nom_prenom',
       'proprietaire_nom_adresse', 'chef_date_naissance_1', 'epouse_prenom',
       'epouse_date_naissance_1', 'chef_permis', 'enfants_date_naissance_1',
       'enfants_profession', 'enfants_permis',
       'pensionnaires_date_naissance_1', 'pensionnaires_permis',
       'chef_date_naissance_3', 'epouse_date_naissance_3',
       'enfants_date_naissance_3', 'pensionnaires_date_naissance_3'],
      dtype='ob

## Corrections automatiques
Dans un premier temps, nous remplaçons les "id" dans la colonne chef_vocation par la vocation dans la ligne en dessus

In [5]:
df_jobs = pd.read_csv(COMPLETE_RAW_DATA)

df_jobs = df_jobs.fillna("")

# Replace "id" by value in row above in chef_vocation
index_id = df_jobs[df_jobs["chef_vocation"].apply(lambda s: s == "id")].index

print(f"There were {len(index_id)} 'id' in 'chef_vocation'")
for idx in index_id:
    df_jobs.loc[idx, "chef_vocation"] = df_jobs.loc[idx-1,"chef_vocation"]
    
df_jobs

/usr/local/Caskroom/miniconda/base/envs/ada/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3441: DtypeWarning: Columns (1) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


There were 154 'id' in 'chef_vocation'


chef_vocation enfants_profession pensionnaires_condition  annee
0           rocureur                           pensre friere se   1835
1            pintier                                     pensre   1835
2          boulanger                                    ouvrier   1835
3          voiturier                                     penvre   1835
4           marechat                            ouvrier|ouvrier   1835
...              ...                ...                     ...    ...
378721        chay l                  ·                       ·   1898
378722   meme maison                  ·                       ·   1898
378723      monimier                  ·                       ·   1898
378724    ontumineur                  ·                       ·   1898
378725     aman bure                  ·                       ·   1898

[378726 rows x 4 columns]

In [6]:
df_jobs = df_jobs.fillna("").replace('·', "") # Remove default value

# Create list form "|"-separated string of job
df_jobs["chef_vocation"] = df_jobs["chef_vocation"].str.split("|")
df_jobs["enfants_profession"] = df_jobs["enfants_profession"].str.split("|")
df_jobs["pensionnaires_condition"] = df_jobs["pensionnaires_condition"].str.split("|")
df_jobs = df_jobs.fillna("").replace('·', "") # Remove default value that emerge from split

In [7]:
# Merge all profession column into profession
df_jobs["profession"] = df_jobs["chef_vocation"] + df_jobs["enfants_profession"] + df_jobs["pensionnaires_condition"]
# Remove empty profession from list
df_jobs["profession"] = df_jobs["profession"].apply(lambda l: [prof for prof in l if len(prof) != 0])


In [8]:
df_jobs

chef_vocation enfants_profession pensionnaires_condition  annee  \
0          [rocureur]                 []      [pensre friere se]   1835   
1           [pintier]                 []                [pensre]   1835   
2         [boulanger]                 []               [ouvrier]   1835   
3         [voiturier]                 []                [penvre]   1835   
4          [marechat]                 []      [ouvrier, ouvrier]   1835   
...               ...                ...                     ...    ...   
378721       [chay l]                 []                      []   1898   
378722  [meme maison]                 []                      []   1898   
378723     [monimier]                 []                      []   1898   
378724   [ontumineur]                 []                      []   1898   
378725    [aman bure]                 []                      []   1898   

                          profession  
0       [rocureur, pensre friere se]  
1                  [pintier, pensre]  
2               [boulanger, ouvrier]  
3                [voiturier, penvre]  
4       [marechat, ouvrier, ouvrier]  
...                              ...  
378721                      [chay l]  
378722                 [meme maison]  
378723                    [monimier]  
378724                  [ontumineur]  
378725                   [aman bure]  

[378726 rows x 5 columns]

In [9]:
df_jobs.to_csv(ORGANIZED_RAW_DATA, index=False)

In [10]:
from ast import literal_eval

df_jobs = pd.read_csv(ORGANIZED_RAW_DATA)
df_jobs[["chef_vocation", "enfants_profession", "pensionnaires_condition", "profession"]] = df_jobs[["chef_vocation", "enfants_profession", "pensionnaires_condition", "profession"]].applymap(literal_eval)
df_jobs

chef_vocation enfants_profession pensionnaires_condition  annee  \
0          [rocureur]                 []      [pensre friere se]   1835   
1           [pintier]                 []                [pensre]   1835   
2         [boulanger]                 []               [ouvrier]   1835   
3         [voiturier]                 []                [penvre]   1835   
4          [marechat]                 []      [ouvrier, ouvrier]   1835   
...               ...                ...                     ...    ...   
378721       [chay l]                 []                      []   1898   
378722  [meme maison]                 []                      []   1898   
378723     [monimier]                 []                      []   1898   
378724   [ontumineur]                 []                      []   1898   
378725    [aman bure]                 []                      []   1898   

                          profession  
0       [rocureur, pensre friere se]  
1                  [pintier, pensre]  
2               [boulanger, ouvrier]  
3                [voiturier, penvre]  
4       [marechat, ouvrier, ouvrier]  
...                              ...  
378721                      [chay l]  
378722                 [meme maison]  
378723                    [monimier]  
378724                  [ontumineur]  
378725                   [aman bure]  

[378726 rows x 5 columns]

Nous devons maintenant séléctionner les données à corriger manuellement.

In [11]:
df_job_year = df_jobs[["profession", "annee"]].explode("profession").copy()
df_job_year = df_job_year[~df_job_year["profession"].isin({"·", ""})]
df_job_year = df_job_year.dropna()
df_job_year

profession  annee
0               rocureur   1835
0       pensre friere se   1835
1                pintier   1835
1                 pensre   1835
2              boulanger   1835
...                  ...    ...
378721            chay l   1898
378722       meme maison   1898
378723          monimier   1898
378724        ontumineur   1898
378725         aman bure   1898

[641406 rows x 2 columns]

Nous avons choisi de garder les top 100 métiers les plus fréquents par année.

In [12]:
top_per_year = 100
df_job_by_year = df_job_year.groupby("annee")["profession"].apply(list)
df_count_top_by_year = df_job_by_year.apply(lambda l: Counter(l).most_common(top_per_year))

df_top_job_by_year = df_count_top_by_year.apply(lambda l: [prof for prof, count in l])
print("Métiers les plus fréquents par année:")
df_top_job_by_year

Métiers les plus fréquents par année:


annee
1835    [domestique, pensionnaire, ouvrier, dom, journ...
1836    [domestique, pensionnaire, rentiere, proprieta...
1837    [domestique, pensionnaire, rentiere, domestiq,...
1838    [domestique, pensionnaire, rentiere, domestiq,...
1839    [domestique, pensionnaire, rentiere, domestiq,...
                              ...                        
1894    [dque, menagere, rentiere, couturiere, journal...
1895    [dque, menagere, rentiere, couturiere, pensre,...
1896    [dque, menagere, rentiere, couturiere, pensre,...
1897    [domestique, menagere, dque, rentiere, couturi...
1898    [domestique, menagere, pensionnaire, rentiere,...
Name: profession, Length: 64, dtype: object

In [13]:
total_occ = len(df_job_year)
print(f"Après les corrections automatiques nous avons {total_occ} occurrences de métier.")

top_terms = set(df_top_job_by_year.explode())
print(f"Il y a {len(top_terms)} termes distincts dans le top {top_per_year} termes les plus fréquents par année.")

total_kept_occ = sum(df_count_top_by_year.apply(lambda l: sum([count for prof, count in l])))
print(f"Après avoir conservé uniquement les occurrences de ces {len(top_terms)} termes distincts, nous avons {total_kept_occ} occurrences de travail.")
print(f"Cela représente {total_kept_occ / total_occ * 100}% des occurrences totales.")

Après les corrections automatiques nous avons 641406 occurrences de métier.
Il y a 532 termes distincts dans le top 100 termes les plus fréquents par année.
Après avoir conservé uniquement les occurrences de ces 532 termes distincts, nous avons 407416 occurrences de travail.
Cela représente 63.51920624378319% des occurrences totales.


Nous utilisons la liste des métiers corrigés de 1855 fournie par les assistants pour les enlever de la listes des métiers à corriger.

In [14]:
df_jobs_1855 = pd.read_excel(CORRECT_TERMS_1855, index_col=0)
df_jobs_1855["prof_norm"] = df_jobs_1855["prof_norm"].apply(lambda x: x.replace("?", ""))
jobs_1855 = set(df_jobs_1855["prof_norm"])
print(f"There is {len(jobs_1855)} corrected jobs in 1885")
jobs_1855

There is 513 corrected jobs in 1885


{"agent d'affaires",
 'agent de change',
 'agriculteur',
 'agricultrice',
 'antiquaire',
 'apprenti',
 'apprentie',
 'architecte',
 'archiviste',
 'armurier',
 'armurrier',
 'arpenteur',
 'artificier',
 'artiste',
 'artiste pedicure',
 'asphalteur',
 'assesseur',
 'aubergiste',
 'avec sa mere',
 'aveugle',
 'avocat',
 'bandagiste',
 'banquier',
 'barbier',
 'batelier',
 'bedeau',
 'belle-mere',
 'bibliothecaire',
 'bijoutier',
 'biscomier',
 'blanchisseuse',
 'boitier',
 'bordeuse',
 'bottier',
 'boucher',
 'boulanger',
 'boulangere',
 'bourrelier',
 'brasseur',
 'brodeuse',
 'brossier',
 'bucheron',
 'buraliste',
 'cafetier',
 'caissier',
 'caissier des postes',
 'capitaine de gendarmerie',
 'carrier',
 'carrossier',
 'cartier',
 'cartonnier',
 'casernier',
 'caviste',
 'chamoiseur',
 'chancelier',
 'chapelain',
 'chapelier',
 'charbonnier',
 'charcutier',
 'charpentier',
 'charpentiere',
 'charretier',
 'charron',
 'chaudronnier',
 'chauffeur',
 "chef d'institut",
 'chez m. renon',
 

In [15]:
already_correct_terms_in_top_100 = top_terms.intersection(jobs_1855)
pd.DataFrame(already_correct_terms_in_top_100).to_csv(ALREADY_CORRECT_TERMS_TOP_100, index=False)

In [16]:
terms_to_correct = top_terms - jobs_1855

print(f"Après avoir supprimé les métiers corrects, il reste {len(terms_to_correct)} termes à vérifier/corriger")

occ_terms = df_job_year[df_job_year["profession"].isin(terms_to_correct)].groupby("profession").count()
occ_terms


Après avoir supprimé les métiers corrects, il reste 385 termes à vérifier/corriger


annee
profession            
agent de police    438
agricult            46
agriculteur id      18
agt                 18
aide               557
...                ...
vintier             65
vocatier            20
vocation          1727
volontaire         264
vve de chambre      52

[385 rows x 1 columns]

In [17]:
occ_terms.sort_values("annee")[::-1]

annee
profession       
manoeuvre    8261
gociant      5818
pens         5402
pensre       3695
gypsier      3000
...           ...
renlier         6
merery          6
echaud          5
ailleuse        5
ordonnier       5

[385 rows x 1 columns]